library/alloc/src/slice.rs

```diff
@@ -179,15 +179,25 @@ impl<T> [T] {
     /// This sort is stable (i.e., does not reorder equal elements) and *O*(*n* \* log(*n*))
     /// worst-case.
     ///
-    /// If `T: Ord` does not implement a total order the resulting order is unspecified. All
-    /// original elements will remain in the slice and any possible modifications via interior
-    /// mutability are observed in the input. Same is true if `T: Ord` panics.
+    /// If the implementation of [`Ord`] for `T` does not implement a [total order] the resulting
+    /// order of elements in the slice is unspecified. All original elements will remain in the
+    /// slice and any possible modifications via interior mutability are observed in the input. Same
```

Comment:
(non-blocking)
Q: Are we promising that we run `Ord::cmp` with each element as an operand at least once, even if there is no total order?

possible A: No, this should be *may* re: mutation, and we are only _really_ promising all elements remain in the slice.

Reply:
The logic of the comparison function is user-defined, so the behavior could depend on some outside sources of non-determinism i.e. network calls, and with that the effects are unbounded.

I'm not sure I fully understand your question. No we don't promise that we run `cmp` for all elements, that alone for the reasons outlined above wouldn't even be enough to have form of "all the logic". Maybe you are referring to the "any possible modifications" part. I'm open to a better formulation here, what I want to communicate is that if your code was executed you will see the effects of that no matter what. The stdlib has been testing this property for a while, and it's required to for a sound implementation. So I think we should document it. Especially the interplay with panics is non-trivial and even some Rust implementations get it wrong [see](https://github.com/Voultapher/sort-research-rs/blob/main/writeup/sort_safety/text.md#property-analysis).